File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> q = new LinkedList<>();
        int currentBridgeWeight = 0;
        int carNumber = 0;
        int elapsedTime = 0;
        
        for(int i = 0 ; i < bridge_length; i++){
            q.add(0);
        }
        
        while(carNumber < truck_weights.length || currentBridgeWeight > 0){
            elapsedTime++;
            
            currentBridgeWeight -= q.poll();
            
            if(carNumber < truck_weights.length && 
               truck_weights[carNumber] + currentBridgeWeight <= weight){
                q.add(truck_weights[carNumber]);
                currentBridgeWeight += truck_weights[carNumber];
                carNumber++;
            } else {
                q.add(0);
            }
        }
        
            return elapsedTime;
        
    }
}

File: JavaScript/프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    const going = [];
    let sum = 0;
    let answer = 0;
    while(truck_weights.length > 0){
        const checkTruck = truck_weights[0];
        if(going.length !== 0){
            for(const e of going){
                e[1]--;
            }
            for(const e of going){
                if(e[1] <= 0){
                    going.shift();
                    sum -= e[0];
                }
            }
        }
        if(going.length < bridge_length && checkTruck <= (weight - sum)){
            const current = truck_weights.shift();
            going.push([current, bridge_length]);
            sum += current;
        }
        answer++;
    }
    return answer + bridge_length;
}